src/main/java/io/bf2/kafka/authorizer/ApiAwareAccessControlEntry.java

```diff
@@ -58,7 +58,11 @@ class ApiAwareAccessControlEntry extends AccessControlEntry {
 
     @Override
     public AclOperation operation() {
-        throw new UnsupportedOperationException("operation");
+        if(!operationsExcluded && operations.size() == 1){
+            return operations.iterator().next();
```

Comment:
This makes me wonder if trying to describe our custom rules via the standard kafka describeAcls API is a good idea.

Our custom rules can represent an include/exclude set of operators. We could possibly explode them into multiple rules just for presentation via describeAcls (so a rule targeting [op1, op2] would be presented as two rules via the api) and potentially the same for an exclusion rule (or we could just ignore exclusions for now as we don't use them for owners).

The custom rules also have the capability to target specific apis, two of our current owner rules target an api set. The only way I can imagine presenting this it is by abusing the principal string. You could construct a new principal during the describe call with additional stuff in it.

I hacked together a quick impl, so it's possible
```
	(principal=User:test_user include [DescribeAcls], host=*, operation=DESCRIBE, permissionType=ALLOW)
	(principal=User:test_user include [CreateAcls,DeleteAcls], host=*, operation=ALTER, permissionType=ALLOW) 
```

We could also potentially mark them as readonly/synthetic this way. Like present the principal as `test_user (readonly)`.

Reply:
abusing the principal like this sounds like we would be doing more harm than good with this PR. maybe we could restrict it so that we only display owner rules that are compatible with the describe response API. So we wouldn't present the rules with api restrictions.

Reply:
@robobario let's not overload the principal.

> Our custom rules can represent an include/exclude set of operators. We could possibly explode them into multiple rules just for presentation via describeAcls (so a rule targeting [op1, op2] would be presented as two rules via the api) and potentially the same for an exclusion rule 

+1 to exploding.

> (or we could just ignore exclusions for now as we don't use them for owners).

+1

> The custom rules also have the capability to target specific apis, two of our current owner rules target an api set. The only way I can imagine presenting this it is by abusing the principal string. You could construct a new principal during the describe call with additional stuff in it.

I wonder whether KIP-482 could offer a solution. Talk to @showuon and see what he thinks of this idea.

https://cwiki.apache.org/confluence/display/KAFKA/KIP-482%3A+The+Kafka+Protocol+should+Support+Optional+Tagged+Fields#KIP482:TheKafkaProtocolshouldSupportOptionalTaggedFields-Motivation

I believe this might be an acceptable use-case as the KIP says:

> Another scenario that we don't support is attaching an extra field to a message in a manner that is orthogonal to the normal versioning scheme.  For example, we might want to attach a trace ID, a "forwarded-by" field, or a "user-agent" field.  It wouldn't make sense to add all these fields to the message schema on the off chance that someone might use them.  In order to support these scenarios, we would like to add the concept of "tagged fields" to the Kafka protocol.

I am wondering if we add a tag `org.bf2.acl.api-list` that carries the list of API.  we could also consider a default-rule tag `org.bf2.acl.default`(value true)  which would allow the UI to known that the rule comes from the default set and it shouldn't offer delete buttons etc.

We'd need to check that the tagged values made it over the REST API.

Reply:
I agree with the point on exploding the multi-operation rules. As you said, the exclusion ones could be done similarly if necessary. KIP-482 looks very interesting @k-wall . I'm interested to see how it's used (haven't found anything yet) and if it will play nicely with clients besides the instance API and CLI that expect the tagged data.

Reply:
KIP-482 cannot work in this case because it needs upstream change (KIP required) to define the request schema to allow tag for a field. 

The definition for `DescribeAclsRequest` is [here](https://github.com/apache/kafka/blob/trunk/clients/src/main/resources/common/message/DescribeAclsRequest.json) (note: no tag field)

A better example is `FetchRequest` [here](https://github.com/apache/kafka/blob/trunk/clients/src/main/resources/common/message/FetchRequest.json#L54-L56). It adds a `clusterId` tagged field for metadata fetch use.